cli/debug.go

```diff
@@ -0,0 +1,108 @@
+	terragruntOptions.Logger.Printf("Debug mode requested: generating debug file")
+
+	variables, err := terraformModuleVariables(terragruntOptions)
+	if err != nil {
+		return err
+	}
+	util.Debugf(terragruntOptions.Logger, "The following variables were detected in the terraform module:")
+	util.Debugf(terragruntOptions.Logger, "%v", variables)
+
+	fileContents, err := terragruntDebugFileContents(terragruntOptions, terragruntConfig, variables)
+	if err != nil {
+		return err
+	}
+
+	configFolder := filepath.Dir(terragruntOptions.TerragruntConfigPath)
+	fileName := filepath.Join(configFolder, "terragrunt-debug.tfvars.json")
+	if err := ioutil.WriteFile(fileName, fileContents, os.FileMode(int(0600))); err != nil {
+		return errors.WithStackTrace(err)
+	}
+
+	terragruntOptions.Logger.Printf("Variables passed to terraform are located in \"%s\"", fileName)
+	terragruntOptions.Logger.Printf("Run this command to replicate how terraform was invoked:")
+	terragruntOptions.Logger.Printf("\tterraform apply -var-file=\"%s\" \"%s\"", fileName, terragruntOptions.WorkingDir)
+	return nil
+}
+
+// terragruntDebugFileContents will return a tfvars file in json format of all the terragrunt rendered variables values
+// that should be set to invoke the terraform module in the same way as terragrunt. Note that this will only include the
+// values of variables that are actually defined in the module.
+func terragruntDebugFileContents(
+	terragruntOptions *options.TerragruntOptions,
+	terragruntConfig *config.TerragruntConfig,
+	moduleVariables []string,
+) ([]byte, error) {
+	var envVars map[string]string
+	if terragruntOptions.Env == nil {
+		envVars = map[string]string{}
+	} else {
+		envVars = terragruntOptions.Env
+	}
+
+	jsonValuesByKey := make(map[string]interface{})
+	for varName, varValue := range terragruntConfig.Inputs {
+		nameAsEnvVar := fmt.Sprintf("TF_VAR_%s", varName)
+		_, varIsInEnv := envVars[nameAsEnvVar]
+		varIsDefined := util.ListContainsElement(moduleVariables, varName)
+
+		// Only add to the file if the explicit env var does NOT exist and the variable is defined in the module.
+		// We must do this in order to avoid overriding the env var when the user follows up with a direct invocation to
+		// terraform using this file (due to the order in which terraform resolves config sources).
+		if !varIsInEnv && varIsDefined {
+			jsonValuesByKey[varName] = varValue
+		} else if varIsInEnv {
+			util.Debugf(
+				terragruntOptions.Logger,
+				"WARN: The variable %s was omitted from the debug file because the env var %s is already set.",
+				varName, nameAsEnvVar,
+			)
+		} else if !varIsDefined {
+			util.Debugf(
+				terragruntOptions.Logger,
+				"WARN: The variable %s was omitted because it is not defined in the terraform module.",
+				varName,
+			)
+		}
+	}
+	jsonContent, err := json.MarshalIndent(jsonValuesByKey, "", "  ")
+	if err != nil {
+		return nil, errors.WithStackTrace(err)
+	}
+	return jsonContent, nil
+}
```

Comment:
This actually produces a usable `tfvars.json` file. This makes me wonder if we should just ALWAYS produce this, and use this to feed the inputs instead of `TF_VAR_` env vars.

Note that I workaround the "vars must all be defined" terraform issue by introspecting the variables from the module source.

The advantage of this is that it would solve all the weird type issues you run into with the env vars (e.g., I think this would allow us to set `null` values for the inputs!).

@brikis98 what do you think?

Reply:
So you're saying that instead of Terragrunt setting env vars to pass `inputs` to a module, we always generate a `tfvars.json` and write it to somewhere in `.terragrunt-cache`? I think I'm in favor of that as it makes the mental model simpler / more visible and helps with type issues, as you mentioned. 

A few things to think through:

1. What should the `.json` file we generate be called? We want to avoid conflicts with files users may already have in their modules. We could do this by either giving it a unique name (e.g., GUID) or by writing it to separate folder from where the user's code is. 
1. How do we tell Terraform to use the file? One option is to name it `terraform.tfvars.json` so Terraform picks it up automatically, but that could clash with the user's own files. Another option is `<GUID>.auto.tfvars.json`. That will avoid file conflicts and be picked up automatically. A third option is to write the file to some other path and explicitly tell Terraform to use it via `-var-file`.
1. I think generating this file would improve interop with TFC/TFE. If we write a `<GUID>.auto.tfvars.json` file, it'll be uploaded to TFC/TFE along with the code during executions.
1. Would we even need a debug mode anymore? I suspect if Terragrunt always logged (a) the path to this file, (b) the exact command its running, and (c) the folder its using as a working dir, then it would be much clearer to users what's going on and no longer require a separate debug command to see it.
1. Longer term thinking: I feel like Terragrunt is becoming a generator for Terraform source code... Here, we're generating a `.tfvars.json` file; with `generate` blocks we also generate `.tf` code; perhaps the [preprocessor idea](https://github.com/gruntwork-io/terragrunt/issues/759#issuecomment-585124357) is also just code generation?

Reply:
There's an issue asking for exactly this, I think... https://github.com/gruntwork-io/terragrunt/issues/752

Reply:
Ok I went ahead and implemented this in a separate PR, so that we can decide which is better: to use this as debug mode, or use this as normal operation. https://github.com/gruntwork-io/terragrunt/pull/1267